File: toolkit/packages/utilities/hooks/statusline.js
#!/usr/bin/env bun

"use strict";

const fs = require("fs");
const { execSync } = require("child_process");
const path = require("path");

// ANSI color constants for terminal coloring
const c = {
    cy: '\033[36m',     // cyan
    g: '\033[32m',      // green
    m: '\033[35m',      // magenta
    gr: '\033[90m',     // gray
    r: '\033[31m',      // red
    o: '\033[38;5;208m', // orange
    y: '\033[33m',      // yellow
    x: '\033[0m'        // reset
};

// Cache for expensive operations
const cache = new Map();
const CACHE_TTL = 1000; // 1 second cache TTL for dynamic data

// Unified execution function with error handling and caching
const exec = (cmd, cwd = null, useCache = false) => {
    const cacheKey = `${cmd}:${cwd || 'default'}`;
    
    if (useCache && cache.has(cacheKey)) {
        const cached = cache.get(cacheKey);
        if (Date.now() - cached.timestamp < CACHE_TTL) {
            return cached.value;
        }
    }
    
    try {
        const options = { 
            encoding: 'utf8', 
            stdio: ['ignore', 'pipe', 'ignore'],
            timeout: 100 // 100ms timeout for git commands
        };
        if (cwd) options.cwd = cwd;
        const result = execSync(cmd, options).trim();
        
        if (useCache) {
            cache.set(cacheKey, { value: result, timestamp: Date.now() });
        }
        
        return result;
    } catch {
        return '';
    }
};

// Fast context percentage calculation - optimized for performance
function getContextPct(transcriptPath) {
    if (!transcriptPath || !fs.existsSync(transcriptPath)) return "0";
    
    try {
        // Use streaming for large files
        const stats = fs.statSync(transcriptPath);
        if (stats.size === 0) return "0";
        
        // Read only last 8KB of file for performance
        const bufferSize = Math.min(8192, stats.size);
        const buffer = Buffer.alloc(bufferSize);
        const fd = fs.openSync(transcriptPath, 'r');
        fs.readSync(fd, buffer, 0, bufferSize, Math.max(0, stats.size - bufferSize));
        fs.closeSync(fd);
        
        const data = buffer.toString('utf8');
        const lines = data.split('\n').filter(l => l.trim());
        
        let latestUsage = null;
        let latestTs = -Infinity;
        
        // Process only last 30 lines for speed
        for (let i = Math.max(0, lines.length - 30); i < lines.length; i++) {
            try {
                const j = JSON.parse(lines[i]);
                const ts = typeof j.timestamp === "string" ? 
                    new Date(j.timestamp).getTime() : j.timestamp;
                const usage = j.message?.usage;
                
                if (ts > latestTs && usage && j.message?.role === "assistant") {
                    latestTs = ts;
                    latestUsage = usage;
                }
            } catch {}
        }
        
        if (latestUsage) {
            const used = (latestUsage.input_tokens || 0) + (latestUsage.output_tokens || 0);
            const limit = latestUsage.cache_creation_input_tokens || 200000;
            const pct = Math.min(100, Math.round((used / limit) * 100));
            return pct.toString();
        }
    } catch {}
    
    return "0";
}

// Extract first user message from transcript
function getFirstUserMessage(transcriptPath) {
    if (!transcriptPath || !fs.existsSync(transcriptPath)) return null;
    
    try {
        const data = fs.readFileSync(transcriptPath, 'utf8');
        const lines = data.split('\n').slice(0, 50); // Check only first 50 lines
        
        for (const line of lines) {
            if (!line.trim()) continue;
            try {
                const j = JSON.parse(line);
                if (j.message?.role === "user" && j.message?.content) {
                    let content = j.message.content;
                    if (Array.isArray(content)) {
                        content = content.find(c => c.type === "text")?.text || "";
                    }
                    return content.slice(0, 100).replace(/\n/g, ' ').trim();
                }
            } catch {}
        }
    } catch {}
    
    return null;
}

// Get or generate session summary with caching
function getSessionSummary(transcriptPath, sessionId, gitDir, workingDir) {
    const summaryDir = path.join(gitDir, '.claude', 'summaries');
    const summaryFile = path.join(summaryDir, `${sessionId}.txt`);
    
    // Check if summary exists
    if (fs.existsSync(summaryFile)) {
        try {
            return fs.readFileSync(summaryFile, 'utf8').trim();
        } catch {}
    }
    
    // Generate new summary from first user message
    const firstMsg = getFirstUserMessage(transcriptPath);
    if (!firstMsg) return null;
    
    // Create summary using git branch context
    const branch = exec('git branch --show-current', workingDir, true) || 'main';
    let summary = firstMsg;
    
    if (branch && branch !== 'main' && branch !== 'master') {
        summary = `[${branch}] ${firstMsg}`;
    }
    
    // Save summary for future use
    try {
        if (!fs.existsSync(summaryDir)) {
            fs.mkdirSync(summaryDir, { recursive: true });
        }
        fs.writeFileSync(summaryFile, summary);
    } catch {}
    
    return summary;
}

// Cached PR lookup with fast git operations
function getPR(branch, workingDir) {
    if (!branch || branch === 'main' || branch === 'master') return '';
    
    const cacheKey = `pr:${branch}`;
    if (cache.has(cacheKey)) {
        const cached = cache.get(cacheKey);
        if (Date.now() - cached.timestamp < 5000) { // 5s cache for PR info
            return cached.value;
        }
    }
    
    // Try to get PR from push output or gh cli
    const prUrl = exec(`git config branch.${branch}.pr-url`, workingDir, true);
    if (prUrl) {
        const prNumber = prUrl.match(/\/pull\/(\d+)/)?.[1];
        if (prNumber) {
            const result = ` ${c.cy}PR#${prNumber}${c.x}`;
            cache.set(cacheKey, { value: result, timestamp: Date.now() });
            return result;
        }
    }
    
    return '';
}

// Main statusline function - optimized for speed
function statusline() {
    try {
        // Parse input JSON
        const input = JSON.parse(fs.readFileSync(0, 'utf8'));
        const workingDir = input.cwd || process.cwd();
        const model = input.model || 'claude-3-5-sonnet';
        const sessionId = input.session_id || 'default';
        const transcriptPath = input.transcript_path;
        
        // Quick model name extraction
        const modelShort = model.includes('opus') ? 'Opus' :
                          model.includes('sonnet') ? 'Sonnet' :
                          model.includes('haiku') ? 'Haiku' : 'Claude';
        
        // Fast git checks
        const gitDir = exec('git rev-parse --show-toplevel', workingDir, true);
        if (!gitDir) {
            // Not in git repo - simple status
            const dir = path.basename(workingDir);
            return `${c.cy}${dir}${c.x} | ${c.m}${modelShort}${c.x}`;
        }
        
        // Get git info with parallel execution where possible
        const branch = exec('git branch --show-current', workingDir, true) || 'HEAD';
        const isWorktree = exec('git rev-parse --is-inside-work-tree', workingDir) === 'true' &&
                          exec('git rev-parse --show-superproject-working-tree', workingDir) === '';
        
        // Fast git status check
        const statusOutput = exec('git status --porcelain', workingDir, true);
        const changes = statusOutput ? statusOutput.split('\n').length : 0;
        
        // Build status components
        const components = [];
        
        // Directory/worktree
        const dirName = path.basename(gitDir);
        if (isWorktree) {
            const worktreeName = path.basename(workingDir);
            components.push(`${c.gr}${dirName}${c.x}/${c.cy}${worktreeName}${c.x}`);
        } else {
            components.push(`${c.cy}${dirName}${c.x}`);
        }
        
        // Branch with color based on name
        const branchColor = branch === 'main' || branch === 'master' ? c.g :
                           branch.startsWith('feature/') ? c.cy :
                           branch.startsWith('fix/') ? c.o :
                           branch.startsWith('hotfix/') ? c.r : c.m;
        components.push(`${branchColor}${branch}${c.x}`);
        
        // Git changes
        if (changes > 0) {
            const changeColor = changes > 10 ? c.r : changes > 5 ? c.y : c.g;
            components.push(`${changeColor}+${changes}${c.x}`);
        }
        
        // PR info
        const pr = getPR(branch, workingDir);
        if (pr) components.push(pr);
        
        // Model and context
        const contextPct = getContextPct(transcriptPath);
        const contextNum = parseInt(contextPct);
        const contextColor = contextNum > 80 ? c.r : contextNum > 60 ? c.y : c.g;
        components.push(`${c.m}${modelShort}${c.x} ${contextColor}${contextPct}%${c.x}`);
        
        // Session summary (only if significant)
        if (contextNum > 10) {
            const summary = getSessionSummary(transcriptPath, sessionId, gitDir, workingDir);
            if (summary && summary.length > 10) {
                const truncated = summary.length > 50 ? 
                    summary.slice(0, 47) + '...' : summary;
                components.push(`${c.gr}# ${truncated}${c.x}`);
            }
        }
        
        return components.join(' | ');
        
    } catch (err) {
        // Fallback status on error
        return `${c.r}statusline error${c.x}`;
    }
}

// Output result immediately
process.stdout.write(statusline());

File: toolkit/packages/utilities/hooks/statusline.py
#!/usr/bin/env python3
"""
ABOUTME: Statusline script for Claude Code showing project, git branch, changes, model, and context
Simple, reliable Python implementation that handles all input cases gracefully
"""

import sys
import json
import os
import subprocess
from pathlib import Path

# ANSI color codes
COLORS = {
    'cyan': '\033[36m',
    'green': '\033[32m',
    'magenta': '\033[35m',
    'gray': '\033[90m',
    'red': '\033[31m',
    'orange': '\033[38;5;208m',
    'yellow': '\033[33m',
    'reset': '\033[0m'
}

def run_git_command(cmd, cwd=None):
    """Run a git command and return output, empty string on error"""
    try:
        result = subprocess.run(
            cmd.split(),
            cwd=cwd or os.getcwd(),
            capture_output=True,
            text=True,
            timeout=0.5
        )
        return result.stdout.strip() if result.returncode == 0 else ''
    except:
        return ''

def get_context_tokens(transcript_path):
    """Calculate actual token usage from transcript"""
    if not transcript_path or not Path(transcript_path).exists():
        return 0
    
    try:
        # Read last few KB of file for performance
        with open(transcript_path, 'rb') as f:
            f.seek(0, 2)  # Go to end
            file_size = f.tell()
            read_size = min(8192, file_size)
            f.seek(max(0, file_size - read_size))
            data = f.read().decode('utf-8', errors='ignore')
        
        # Find the most recent usage info
        lines = data.strip().split('\n')
        for line in reversed(lines[-30:]):  # Check last 30 lines
            try:
                entry = json.loads(line)
                if 'message' in entry and 'usage' in entry['message']:
                    usage = entry['message']['usage']
                    if entry['message'].get('role') == 'assistant':
                        input_tokens = usage.get('input_tokens', 0)
                        output_tokens = usage.get('output_tokens', 0)
                        total_tokens = input_tokens + output_tokens
                        return total_tokens
            except:
                continue
    except:
        pass
    
    return 0

def get_model_short_name(model_info):
    """Extract short model name from model object or string"""
    # Handle both object format and string format for backward compatibility
    if isinstance(model_info, dict):
        model_name = model_info.get('display_name', model_info.get('id', 'Claude'))
    else:
        model_name = str(model_info)
    
    if 'opus' in model_name.lower():
        return 'Opus'
    elif 'sonnet' in model_name.lower():
        return 'Sonnet' 
    elif 'haiku' in model_name.lower():
        return 'Haiku'
    return 'Claude'

def format_token_count(tokens):
    """Format token count for display (e.g. 12500 -> '12.5k')"""
    if tokens == 0:
        return '0'
    elif tokens < 1000:
        return str(tokens)
    elif tokens < 1000000:
        return f"{tokens/1000:.1f}k"
    else:
        return f"{tokens/1000000:.1f}M"

def replace_home_with_tilde(path):
    """Replace home directory with ~ in path"""
    home = os.path.expanduser('~')
    if path.startswith(home):
        return path.replace(home, '~', 1)
    return path

def build_statusline():
    """Build the statusline string"""
    # Parse input - handle all cases
    input_data = {}
    try:
        raw_input = sys.stdin.read()
        if raw_input.strip():
            input_data = json.loads(raw_input)
    except:
        pass  # Use defaults
    
    # Extract values with defaults
    cwd = input_data.get('cwd', os.getcwd())
    model = input_data.get('model', 'claude-3-5-sonnet')
    session_id = input_data.get('session_id', 'default')
    transcript_path = input_data.get('transcript_path', '')
    
    # Get short model name (handles both object and string format)
    model_short = get_model_short_name(model)
    
    # Replace home directory with tilde in current directory
    display_cwd = replace_home_with_tilde(cwd)
    
    # Check if we're in a git repo
    git_root = run_git_command('git rev-parse --show-toplevel', cwd)
    
    if not git_root:
        # Not in git repo - simple status with current directory and model
        dir_name = Path(display_cwd).name
        tokens = get_context_tokens(transcript_path)
        token_display = format_token_count(tokens)
        
        components = []
        components.append(f"{COLORS['cyan']}{dir_name}{COLORS['reset']}")
        components.append(f"{COLORS['orange']}{model_short}{COLORS['reset']}")
        if tokens > 0:
            components.append(f"{COLORS['gray']}{token_display} tokens{COLORS['reset']}")
        
        return ' | '.join(components)
    
    # Get git information
    branch = run_git_command('git branch --show-current', cwd) or 'HEAD'
    
    # Count git changes
    status_output = run_git_command('git status --porcelain', cwd)
    changes_count = len(status_output.split('\n')) if status_output else 0
    
    # Build components
    components = []
    
    # Project directory
    project_name = Path(git_root).name
    components.append(f"{COLORS['cyan']}{project_name}{COLORS['reset']}")
    
    # Git branch with color
    if branch in ['main', 'master']:
        branch_color = COLORS['green']
    elif branch.startswith('feature/'):
        branch_color = COLORS['cyan']
    elif branch.startswith('fix/'):
        branch_color = COLORS['orange']
    elif branch.startswith('hotfix/'):
        branch_color = COLORS['red']
    else:
        branch_color = COLORS['magenta']
    
    components.append(f"{branch_color}{branch}{COLORS['reset']}")
    
    # Git changes
    if changes_count > 0:
        if changes_count > 10:
            change_color = COLORS['red']
        elif changes_count > 5:
            change_color = COLORS['yellow']
        else:
            change_color = COLORS['green']
        components.append(f"{change_color}+{changes_count}{COLORS['reset']}")
    
    # Model and token usage
    tokens = get_context_tokens(transcript_path)
    token_display = format_token_count(tokens)
    
    # Color based on token usage (rough thresholds)
    if tokens > 150000:
        token_color = COLORS['red']
    elif tokens > 100000:
        token_color = COLORS['yellow']
    else:
        token_color = COLORS['green']
    
    model_component = f"{COLORS['orange']}{model_short}{COLORS['reset']}"
    if tokens > 0:
        model_component += f" {token_color}{token_display} tokens{COLORS['reset']}"
    
    components.append(model_component)
    
    # Session summary (optional - if token usage > 10k)
    if tokens > 10000 and transcript_path and Path(transcript_path).exists():
        try:
            # Try to get first user message for summary
            with open(transcript_path, 'r') as f:
                for line in f:
                    try:
                        entry = json.loads(line)
                        if entry.get('message', {}).get('role') == 'user':
                            content = entry['message'].get('content', '')
                            if isinstance(content, list):
                                for item in content:
                                    if item.get('type') == 'text':
                                        content = item.get('text', '')
                                        break
                            if content:
                                summary = content[:50].replace('\n', ' ').strip()
                                if len(content) > 50:
                                    summary += '...'
                                components.append(f"{COLORS['gray']}# {summary}{COLORS['reset']}")
                                break
                    except:
                        continue
        except:
            pass
    
    return ' | '.join(components)

if __name__ == '__main__':
    try:
        statusline = build_statusline()
        sys.stdout.write(statusline)
        sys.stdout.flush()
    except Exception as e:
        # On any error, show a simple fallback
        sys.stdout.write(f"{COLORS['red']}status error{COLORS['reset']}")
        sys.stdout.flush()
